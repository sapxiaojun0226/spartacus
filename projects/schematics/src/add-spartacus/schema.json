{
  "$schema": "http://json-schema.org/schema",
  "$id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": [
          "ASM",
          "Import-Export",
          "Saved-Cart",
          "Quick-Order",
          "CDC",
          "CDS",
          "Cart",
          "WishList",
          "Checkout",
          "Checkout-B2B",
          "Checkout-Scheduled-Replenishment",
          "Order",
          "Digital-Payments",
          "EPD-Visualization",
          "Administration",
          "Order-Approval",
          "Account-Summary",
          "Bulk-Pricing",
          "Image-Zoom",
          "Product-Variants",
          "VC-Configurator",
          "Textfield-Configurator",
          "CPQ-Configurator",
          "Qualtrics",
          "SmartEdit",
          "Store-Finder",
          "Personalization",
          "TMS-GTM",
          "TMS-AEPL",
          "User-Account",
          "User-Profile"
        ],
        "type": "string"
      },
      "default": [
        "ASM",
        "Cart",
        "WishList",
        "Checkout",
        "Order",
        "Import-Export",
        "Saved-Cart",
        "Quick-Order",
        "Image-Zoom",
        "Product-Variants",
        "SmartEdit",
        "Store-Finder",
        "Personalization",
        "User-Account",
        "User-Profile"
      ],
      "x-prompt": {
        "message": "Which Spartacus features would you like to set up?\nPlease note that for most Spartacus features to be properly configured, the User-Account feature is required.",
        "type": "list",
        "items": [
          {
            "value": "ASM",
            "label": "Assisted Services Module"
          },
          {
            "value": "Import-Export",
            "label": "Import/Export"
          },
          {
            "value": "Saved-Cart",
            "label": "Saved Cart"
          },
          {
            "value": "Quick-Order",
            "label": "Quick Order"
          },
          {
            "value": "CDC",
            "label": "Customer Data Cloud Integration"
          },
          {
            "value": "CDS",
            "label": "Context-Driven Services Integration"
          },
          {
            "value": "Checkout",
            "label": "Checkout base"
          },
          {
            "value": "Checkout-B2B",
            "label": "Checkout B2B (b2b feature, requires Base Checkout)"
          },
          {
            "value": "Checkout-Scheduled-Replenishment",
            "label": "Checkout Scheduled Replenishment (b2b feature, requires Base and B2B Checkout)"
          },
          {
            "value": "Cart",
            "label": "Cart"
          },
          {
            "value": "WishList",
            "label": "WishList"
          },
          {
            "value": "Order",
            "label": "Order"
          },
          {
            "value": "Digital-Payments",
            "label": "Digital Payments Integration"
          },
          {
            "value": "EPD-Visualization",
            "label": "EPD Visualization Integration"
          },
          {
            "value": "Administration",
            "label": "Organization - Adminstration (b2b feature)"
          },
          {
            "value": "Order-Approval",
            "label": "Organization - Order Approval (b2b feature)"
          },
          {
            "value": "Account-Summary",
            "label": "Organization - Account Summary (b2b feature, requires Organization - Administration)"
          },
          {
            "value": "Bulk-Pricing",
            "label": "Product - Bulk Pricing (b2b feature)"
          },
          {
            "value": "Product-Variants",
            "label": "Product - Variants"
          },
          {
            "value": "Image-Zoom",
            "label": "Product - Image Zoom"
          },
          {
            "value": "VC-Configurator",
            "label": "Product Configurator - Variant Configurator"
          },
          {
            "value": "Textfield-Configurator",
            "label": "Product Configurator - Textfield Configurator"
          },
          {
            "value": "CPQ-Configurator",
            "label": "Product Configurator - CPQ Configurator (b2b feature)"
          },
          {
            "value": "Qualtrics",
            "label": "Qualtrics"
          },
          {
            "value": "SmartEdit",
            "label": "SmartEdit"
          },
          {
            "value": "Store-Finder",
            "label": "Store Finder"
          },
          {
            "value": "Personalization",
            "label": "Tracking - Personalization"
          },
          {
            "value": "TMS-GTM",
            "label": "Tracking - Tag Management System - Google Tag Manager"
          },
          {
            "value": "TMS-AEPL",
            "label": "Tracking - Tag Management System - Adobe Experience Platform Launch"
          },
          {
            "value": "User-Account",
            "label": "User - Account"
          },
          {
            "value": "User-Profile",
            "label": "User - Profile"
          }
        ]
      }
    },
    "debug": {
      "description": "Display additional details during the running process.",
      "type": "boolean",
      "default": false
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the Spartacus features.",
      "default": true
    },
    "baseUrl": {
      "type": "string",
      "description": "The url of the OCC backend",
      "default": "https://localhost:9002"
    },
    "occPrefix": {
      "type": "string",
      "description": "The OCC API prefix. E.g.: /occ/v2/"
    },
    "baseSite": {
      "type": "string",
      "description": "A comma-separated list of base site(s) to use with Spartacus"
    },
    "currency": {
      "type": "string",
      "description": "A comma-separated list of currencies to use in Spartacus"
    },
    "language": {
      "type": "string",
      "description": "A comma-separated list of languages to use in Spartacus"
    },
    "urlParameters": {
      "type": "string",
      "description": "A comma-separated list that represents the order of site-context parameters. E.g.: ['baseSite', 'language', 'currency']"
    },
    "useMetaTags": {
      "type": "boolean",
      "description": "Configure baseUrl via meta tag in index.html"
    },
    "featureLevel": {
      "type": "string",
      "description": "Configure feature level for application."
    },
    "overwriteAppComponent": {
      "type": "boolean",
      "description": "Overwrite the contents of app.component.html file with the <cx-storefront> tag.",
      "default": true
    },
    "pwa": {
      "type": "boolean",
      "description": "Add PWA support to the project.",
      "default": false
    },
    "ssr": {
      "type": "boolean",
      "description": "Add Server-side rendering support to the project.",
      "default": false
    },
    "theme": {
      "type": "string",
      "description": "Select a style theme to add. E.g: santorini"
    }
  },
  "required": []
}
