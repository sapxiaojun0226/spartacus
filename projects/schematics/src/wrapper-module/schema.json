{
  "$schema": "http://json-schema.org/schema",
  "$id": "SpartacusSchematics",
  "title": "Spartacus Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "debug": {
      "description": "Display additional details during the running process.",
      "type": "boolean",
      "default": false
    },
    "markerModuleName": {
      "type": "string",
      "description": "The marker module for which to find or generate the wrapper module."
    },
    "featureModuleName": {
      "type": "string",
      "description": "Name of the feature module which to add to the wrapper module."
    }
  },
  "required": ["markerModuleName", "featureModuleName"]
}
