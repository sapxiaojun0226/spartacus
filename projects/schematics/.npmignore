# Ignores TypeScript files, but keeps definitions.
*.ts
!*.d.ts
yarn.lock
*_spec.ts
*_spec.ts.snap
test-jest.ts
jest.schematics.config.js
tsconfig.schematics.json
src/migrations/test/**/*
.eslintrc.json
sub-folder/
.pipeline/
