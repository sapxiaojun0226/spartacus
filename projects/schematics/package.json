{
  "name": "@spartacus/schematics",
  "version": "4.1.0-next.0",
  "description": "Spartacus schematics",
  "keywords": [
    "spartacus",
    "schematics"
  ],
  "homepage": "https://github.com/SAP/spartacus",
  "repository": "https://github.com/SAP/spartacus/tree/develop/projects/schematics",
  "license": "MIT",
  "main": "index.js",
  "scripts": {
    "build": "yarn clean && yarn --cwd ../../ generate:deps && ../../node_modules/.bin/tsc -p ./tsconfig.schematics.json",
    "clean": "../../node_modules/.bin/rimraf \"index.js*\" \"index.d.ts\" \"src/**/*.js\" \"src/**/*.js.map\" \"src/**/*.d.ts\"",
    "test": "yarn clean && ../../node_modules/.bin/jest --config ./jest.schematics.config.js"
  },
  "dependencies": {
    "@angular/pwa": "^14.2.3",
    "@nguniversal/express-engine": "^14.0.0",
    "semver": "^7.3.5",
    "ts-morph": "^9.1.0",
    "tslib": "^2.4.0"
  },
  "peerDependencies": {
    "@angular-devkit/core": "^14.2.3",
    "@angular-devkit/schematics": "^14.2.3",
    "@angular/compiler": "^14.2.3",
    "@schematics/angular": "^14.2.3",
    "jsonc-parser": "^3.0.0",
    "parse5": "^6.0.1",
    "typescript": "^4.7.2"
  },
  "peerDependenciesMeta": {
    "canvas": {
      "optional": true
    }
  },
  "ng-add": {
    "save": "devDependencies"
  },
  "ng-update": {
    "migrations": "./src/migrations/migrations.json",
    "packageGroup": [
      "@spartacus/core",
      "@spartacus/storefront",
      "@spartacus/styles",
      "@spartacus/assets",
      "@spartacus/setup",
      "@spartacus/asm",
      "@spartacus/cart",
      "@spartacus/checkout",
      "@spartacus/order",
      "@spartacus/organization",
      "@spartacus/product",
      "@spartacus/product-configurator",
      "@spartacus/qualtrics",
      "@spartacus/smartedit",
      "@spartacus/storefinder",
      "@spartacus/tracking",
      "@spartacus/user",
      "@spartacus/cdc",
      "@spartacus/cds",
      "@spartacus/digital-payments",
      "@spartacus/epd-visualization"
    ]
  },
  "schematics": "./src/collection.json"
}
