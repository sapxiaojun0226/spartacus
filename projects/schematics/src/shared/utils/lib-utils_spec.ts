import { Tree } from '@angular-devkit/schematics';
import { SchematicTestRunner } from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import * as path from 'path';
import { Schema as SpartacusOptions } from '../../add-spartacus/schema';
import { CDS_CONFIG, UTF_8 } from '../constants';
import {
  SPARTACUS_CART,
  SPARTACUS_CDS,
  SPARTACUS_CHECKOUT,
  SPARTACUS_ORDER,
} from '../libs-constants';
import {
  addLibraryFeature,
  addPackageJsonDependenciesForLibrary,
  LibraryOptions,
  SchematicConfig,
  shouldAddFeature,
} from './lib-utils';

const xxxFeaturePath = `src/app/spartacus/features/xxx/xxx-feature.module.ts`;

describe('Lib utils', () => {
  const schematicRunner = new SchematicTestRunner(
    'schematics',
    path.join(__dirname, '../../collection.json')
  );

  let appTree: Tree;

  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };

  const spartacusDefaultOptions: SpartacusOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [],
  };

  const CLI_FEATURE_NAME = 'xxx-cli';
  const FEATURE_NAME = 'xxx';
  const FEATURE_FOLDER_NAME = 'xxx';
  const FEATURE_MODULE_NAME = 'XxxModule';
  const FEATURE_MODULE_IMPORT_PATH = '@spartacus/xxx';
  const ROOT_MODULE_NAME = 'XxxModuleRoot';
  const ROOT_FEATURE_MODULE_IMPORT_PATH = '@spartacus/xxx/root';
  const I18N_RESOURCES = 'translations';
  const I18N_CHUNKS = 'translationChunk';
  const ASSETS_IMPORT_PATH = '@spartacus/xxx/assets';
  const SCSS_FILE_NAME = 'xxx.scss';
  const STYLE_IMPORT_PATH = FEATURE_MODULE_IMPORT_PATH;

  const scssFilePath = `src/styles/spartacus/${SCSS_FILE_NAME}`;

  const BASE_FEATURE_CONFIG: SchematicConfig = {
    library: {
      featureName: CLI_FEATURE_NAME,
      mainScope: FEATURE_MODULE_IMPORT_PATH,
    },
    folderName: FEATURE_FOLDER_NAME,
    moduleName: FEATURE_NAME,
    featureModule: {
      name: FEATURE_MODULE_NAME,
      importPath: FEATURE_MODULE_IMPORT_PATH,
    },
    rootModule: {
      name: ROOT_MODULE_NAME,
      importPath: ROOT_FEATURE_MODULE_IMPORT_PATH,
    },
    i18n: {
      resources: I18N_RESOURCES,
      chunks: I18N_CHUNKS,
      importPath: ASSETS_IMPORT_PATH,
    },
    styles: {
      scssFileName: SCSS_FILE_NAME,
      importStyle: STYLE_IMPORT_PATH,
    },
  };

  const BASE_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    features: [CLI_FEATURE_NAME],
    lazy: true,
  };

  const CHECKOUT_FEATURE_CONFIG: SchematicConfig = {
    library: {
      featureName: 'checkout',
      mainScope: '@spartacus/checkout',
      featureScope: '@spartacus/checkout/base',
    },
    folderName: 'checkout',
    moduleName: 'Checkout',
    featureModule: {
      name: 'CheckoutModule',
      importPath: '@spartacus/checkout/base',
    },
    rootModule: {
      name: 'CheckoutRootModule',
      importPath: '@spartacus/checkout/base/root',
    },
  };
  const CHECKOUT_OPTIONS: LibraryOptions = {
    project: 'schematics-test',
    features: ['checkout-base-cli'],
    lazy: true,
  };

  beforeEach(async () => {
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runSchematicAsync(
        'add-spartacus',
        { ...spartacusDefaultOptions, name: 'schematics-test' },
        appTree
      )
      .toPromise();
  });

  describe('shouldAddFeature', () => {
    it('should return true if the feature is present in the given features array', () => {
      const feature1 = 'feature1';
      const features = [feature1];
      expect(shouldAddFeature(feature1, features)).toBe(true);
    });
    it('should return false if the feature is NOT present in the given features array', () => {
      const random = 'random';
      const feature1 = 'feature1';
      const features = [feature1];
      expect(shouldAddFeature(random, features)).toBe(false);
    });
  });

  describe('addLibraryFeature', () => {
    it('should add i18n config in feature module', async () => {
      appTree = await schematicRunner
        .callRule(addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG), appTree)
        .toPromise();

      expect(appTree.read(xxxFeaturePath)?.toString(UTF_8)).toMatchSnapshot();
    });
    it('should NOT add i18n if the config is not present', async () => {
      const featureConfig: SchematicConfig = {
        ...BASE_FEATURE_CONFIG,
        i18n: undefined,
      };
      appTree = await schematicRunner
        .callRule(addLibraryFeature(BASE_OPTIONS, featureConfig), appTree)
        .toPromise();

      expect(appTree.read(xxxFeaturePath)?.toString(UTF_8)).toMatchSnapshot();
    });
    describe('when the lazy loading is configured', () => {
      it('should add it in the lazy loading way', async () => {
        appTree = await schematicRunner
          .callRule(
            addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG),
            appTree
          )
          .toPromise();

        expect(appTree.read(xxxFeaturePath)?.toString(UTF_8)).toMatchSnapshot();
      });
    });
    describe('when the eager loading is configured', () => {
      it('should add it in the eager way', async () => {
        const rule = addLibraryFeature(
          { ...BASE_OPTIONS, lazy: false },
          BASE_FEATURE_CONFIG
        );
        appTree = await schematicRunner.callRule(rule, appTree).toPromise();

        expect(appTree.read(xxxFeaturePath)?.toString(UTF_8)).toMatchSnapshot();
      });
    });
    describe('custom config option', () => {
      it('should add the custom config when set', async () => {
        const featureConfig: SchematicConfig = {
          ...BASE_FEATURE_CONFIG,
          customConfig: () => ({
            providers: {
              import: [
                {
                  moduleSpecifier: SPARTACUS_CDS,
                  namedImports: [CDS_CONFIG],
                },
              ],
              content: `<${CDS_CONFIG}>{
              cds: {
                profileTag: {
                  javascriptUrl:
                    'random',
                  configUrl:
                    'random',
                  allowInsecureCookies: true,
                },
              },
            }`,
            },
          }),
        };
        appTree = await schematicRunner
          .callRule(addLibraryFeature(BASE_OPTIONS, featureConfig), appTree)
          .toPromise();

        expect(appTree.read(xxxFeaturePath)?.toString(UTF_8)).toMatchSnapshot();
      });
    });
    describe('assets options', () => {
      it('should update angular.json file with assets', async () => {
        // before
        expect(appTree.read('angular.json')?.toString(UTF_8)).toMatchSnapshot();

        const featureConfig: SchematicConfig = {
          ...BASE_FEATURE_CONFIG,
          assets: {
            input: 'smartedit/assets',
            glob: '**/*',
          },
        };
        appTree = await schematicRunner
          .callRule(addLibraryFeature(BASE_OPTIONS, featureConfig), appTree)
          .toPromise();

        // after
        expect(appTree.read('angular.json')?.toString(UTF_8)).toMatchSnapshot();
      });
    });
    describe('style', () => {
      describe('when style config is provided', () => {
        describe('and the scss file does NOT exist', () => {
          it('should add it', async () => {
            const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
            appTree = await schematicRunner.callRule(rule, appTree).toPromise();

            expect(appTree.exists(scssFilePath)).toEqual(true);
            const content = appTree.read(scssFilePath)?.toString(UTF_8);
            expect(content).toEqual(`@import "${FEATURE_MODULE_IMPORT_PATH}";`);
          });
        });
        describe('and the scss with the same content already exists', () => {
          beforeEach(() => {
            appTree.create(
              scssFilePath,
              `@import "${FEATURE_MODULE_IMPORT_PATH}";`
            );
          });
          it('should NOT append it', async () => {
            const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
            appTree = await schematicRunner.callRule(rule, appTree).toPromise();

            expect(appTree.exists(scssFilePath)).toEqual(true);
            const content = appTree.read(scssFilePath)?.toString(UTF_8);
            expect(content).toEqual(`@import "${FEATURE_MODULE_IMPORT_PATH}";`);
          });
        });
        describe('and the scss file with a different content already exists', () => {
          const randomContent = `@import "@random/xxx";`;
          beforeEach(() => {
            appTree.create(scssFilePath, randomContent);
          });
          it('should append it', async () => {
            const rule = addLibraryFeature(BASE_OPTIONS, BASE_FEATURE_CONFIG);
            appTree = await schematicRunner.callRule(rule, appTree).toPromise();

            expect(appTree.exists(scssFilePath)).toEqual(true);
            const content = appTree.read(scssFilePath)?.toString(UTF_8);
            expect(content).toEqual(
              `${randomContent}\n@import "${FEATURE_MODULE_IMPORT_PATH}";`
            );
          });
        });
      });
      describe('when style config is NOT provided', () => {
        it('should not add it', async () => {
          const rule = addLibraryFeature(BASE_OPTIONS, {
            ...BASE_FEATURE_CONFIG,
            styles: undefined,
          });
          appTree = await schematicRunner.callRule(rule, appTree).toPromise();

          expect(appTree.exists(scssFilePath)).toEqual(false);
        });
      });
    });
  });

  describe('addPackageJsonDependenciesForLibrary', () => {
    beforeEach(async () => {
      appTree = await schematicRunner
        .callRule(
          addLibraryFeature(CHECKOUT_OPTIONS, CHECKOUT_FEATURE_CONFIG),
          appTree
        )
        .toPromise();
    });

    it('checkout', async () => {
      const peerDependencies: Record<string, string> = {
        [SPARTACUS_ORDER]: '4.1.0-next.0',
        [SPARTACUS_CART]: '4.1.0-next.0',
        [SPARTACUS_CHECKOUT]: '4.1.0-next.0',
      };

      await schematicRunner
        .callRule(
          addPackageJsonDependenciesForLibrary(
            peerDependencies,
            CHECKOUT_OPTIONS
          ),
          appTree
        )
        .toPromise();

      const packageJson = JSON.parse(
        appTree.read('package.json')?.toString(UTF_8) ?? ''
      ).dependencies as Record<string, string>;

      expect(packageJson[SPARTACUS_ORDER]).toBeTruthy();
      expect(packageJson[SPARTACUS_CART]).toBeTruthy();
      expect(packageJson[SPARTACUS_CHECKOUT]).toBeTruthy();
    });
  });
});
