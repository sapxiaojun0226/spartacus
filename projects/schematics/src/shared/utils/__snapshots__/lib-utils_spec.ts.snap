// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lib utils addLibraryFeature assets options should update angular.json file with assets 1`] = `
"{
  \\"$schema\\": \\"./node_modules/@angular/cli/lib/config/schema.json\\",
  \\"version\\": 1,
  \\"newProjectRoot\\": \\"\\",
  \\"projects\\": {
    \\"schematics-test\\": {
      \\"projectType\\": \\"application\\",
      \\"schematics\\": {
        \\"@schematics/angular:component\\": {
          \\"style\\": \\"scss\\"
        }
      },
      \\"root\\": \\"\\",
      \\"sourceRoot\\": \\"src\\",
      \\"prefix\\": \\"app\\",
      \\"architect\\": {
        \\"build\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:browser\\",
          \\"options\\": {
            \\"outputPath\\": \\"dist/schematics-test\\",
            \\"index\\": \\"src/index.html\\",
            \\"main\\": \\"src/main.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.app.json\\",
            \\"inlineStyleLanguage\\": \\"scss\\",
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\"
            ],
            \\"styles\\": [
              \\"src/styles.scss\\"
            ],
            \\"scripts\\": [],
            \\"stylePreprocessorOptions\\": {
              \\"includePaths\\": [
                \\"node_modules/\\"
              ]
            }
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"budgets\\": [
                {
                  \\"type\\": \\"initial\\",
                  \\"maximumWarning\\": \\"500kb\\",
                  \\"maximumError\\": \\"2.5mb\\"
                },
                {
                  \\"type\\": \\"anyComponentStyle\\",
                  \\"maximumWarning\\": \\"2kb\\",
                  \\"maximumError\\": \\"4kb\\"
                }
              ],
              \\"fileReplacements\\": [
                {
                  \\"replace\\": \\"src/environments/environment.ts\\",
                  \\"with\\": \\"src/environments/environment.prod.ts\\"
                }
              ],
              \\"outputHashing\\": \\"all\\"
            },
            \\"development\\": {
              \\"buildOptimizer\\": false,
              \\"optimization\\": false,
              \\"vendorChunk\\": true,
              \\"extractLicenses\\": false,
              \\"sourceMap\\": true,
              \\"namedChunks\\": true
            }
          },
          \\"defaultConfiguration\\": \\"production\\"
        },
        \\"serve\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:dev-server\\",
          \\"configurations\\": {
            \\"production\\": {
              \\"browserTarget\\": \\"schematics-test:build:production\\"
            },
            \\"development\\": {
              \\"browserTarget\\": \\"schematics-test:build:development\\"
            }
          },
          \\"defaultConfiguration\\": \\"development\\"
        },
        \\"extract-i18n\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:extract-i18n\\",
          \\"options\\": {
            \\"browserTarget\\": \\"schematics-test:build\\"
          }
        },
        \\"test\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:karma\\",
          \\"options\\": {
            \\"main\\": \\"src/test.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.spec.json\\",
            \\"karmaConfig\\": \\"karma.conf.js\\",
            \\"inlineStyleLanguage\\": \\"scss\\",
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\"
            ],
            \\"styles\\": [
              \\"src/styles.scss\\"
            ],
            \\"scripts\\": [],
            \\"stylePreprocessorOptions\\": {
              \\"includePaths\\": [
                \\"node_modules/\\"
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`Lib utils addLibraryFeature assets options should update angular.json file with assets 2`] = `
"{
  \\"$schema\\": \\"./node_modules/@angular/cli/lib/config/schema.json\\",
  \\"version\\": 1,
  \\"newProjectRoot\\": \\"\\",
  \\"projects\\": {
    \\"schematics-test\\": {
      \\"projectType\\": \\"application\\",
      \\"schematics\\": {
        \\"@schematics/angular:component\\": {
          \\"style\\": \\"scss\\"
        }
      },
      \\"root\\": \\"\\",
      \\"sourceRoot\\": \\"src\\",
      \\"prefix\\": \\"app\\",
      \\"architect\\": {
        \\"build\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:browser\\",
          \\"options\\": {
            \\"outputPath\\": \\"dist/schematics-test\\",
            \\"index\\": \\"src/index.html\\",
            \\"main\\": \\"src/main.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.app.json\\",
            \\"inlineStyleLanguage\\": \\"scss\\",
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\",
              {
                \\"glob\\": \\"**/*\\",
                \\"input\\": \\"./node_modules/@spartacus/smartedit/assets\\",
                \\"output\\": \\"assets/\\"
              }
            ],
            \\"styles\\": [
              \\"src/styles.scss\\",
              \\"src/styles/spartacus/xxx.scss\\"
            ],
            \\"scripts\\": [],
            \\"stylePreprocessorOptions\\": {
              \\"includePaths\\": [
                \\"node_modules/\\"
              ]
            }
          },
          \\"configurations\\": {
            \\"production\\": {
              \\"budgets\\": [
                {
                  \\"type\\": \\"initial\\",
                  \\"maximumWarning\\": \\"500kb\\",
                  \\"maximumError\\": \\"2.5mb\\"
                },
                {
                  \\"type\\": \\"anyComponentStyle\\",
                  \\"maximumWarning\\": \\"2kb\\",
                  \\"maximumError\\": \\"4kb\\"
                }
              ],
              \\"fileReplacements\\": [
                {
                  \\"replace\\": \\"src/environments/environment.ts\\",
                  \\"with\\": \\"src/environments/environment.prod.ts\\"
                }
              ],
              \\"outputHashing\\": \\"all\\"
            },
            \\"development\\": {
              \\"buildOptimizer\\": false,
              \\"optimization\\": false,
              \\"vendorChunk\\": true,
              \\"extractLicenses\\": false,
              \\"sourceMap\\": true,
              \\"namedChunks\\": true
            }
          },
          \\"defaultConfiguration\\": \\"production\\"
        },
        \\"serve\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:dev-server\\",
          \\"configurations\\": {
            \\"production\\": {
              \\"browserTarget\\": \\"schematics-test:build:production\\"
            },
            \\"development\\": {
              \\"browserTarget\\": \\"schematics-test:build:development\\"
            }
          },
          \\"defaultConfiguration\\": \\"development\\"
        },
        \\"extract-i18n\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:extract-i18n\\",
          \\"options\\": {
            \\"browserTarget\\": \\"schematics-test:build\\"
          }
        },
        \\"test\\": {
          \\"builder\\": \\"@angular-devkit/build-angular:karma\\",
          \\"options\\": {
            \\"main\\": \\"src/test.ts\\",
            \\"polyfills\\": \\"src/polyfills.ts\\",
            \\"tsConfig\\": \\"tsconfig.spec.json\\",
            \\"karmaConfig\\": \\"karma.conf.js\\",
            \\"inlineStyleLanguage\\": \\"scss\\",
            \\"assets\\": [
              \\"src/favicon.ico\\",
              \\"src/assets\\",
              {
                \\"glob\\": \\"**/*\\",
                \\"input\\": \\"./node_modules/@spartacus/smartedit/assets\\",
                \\"output\\": \\"assets/\\"
              }
            ],
            \\"styles\\": [
              \\"src/styles.scss\\",
              \\"src/styles/spartacus/xxx.scss\\"
            ],
            \\"scripts\\": [],
            \\"stylePreprocessorOptions\\": {
              \\"includePaths\\": [
                \\"node_modules/\\"
              ]
            }
          }
        }
      }
    }
  }
}"
`;

exports[`Lib utils addLibraryFeature custom config option should add the custom config when set 1`] = `
"import { NgModule } from '@angular/core';
import { CdsConfig } from \\"@spartacus/cds\\";
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { translationChunk, translations } from \\"@spartacus/xxx/assets\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      xxx: {
        module: () =>
          import('@spartacus/xxx').then((m) => m.XxxModule),
      },
    }
  }),
  provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunk,
    },
  }),
  provideConfig(<CdsConfig>{
    cds: {
      profileTag: {
        javascriptUrl:
          'random',
        configUrl:
          'random',
        allowInsecureCookies: true,
      },
    },
  })
  ]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils addLibraryFeature should NOT add i18n if the config is not present 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, provideConfig } from \\"@spartacus/core\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      xxx: {
        module: () =>
          import('@spartacus/xxx').then((m) => m.XxxModule),
      },
    }
  })]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils addLibraryFeature should add i18n config in feature module 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { translationChunk, translations } from \\"@spartacus/xxx/assets\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      xxx: {
        module: () =>
          import('@spartacus/xxx').then((m) => m.XxxModule),
      },
    }
  }),
  provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunk,
    },
  })
  ]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils addLibraryFeature when the eager loading is configured should add it in the eager way 1`] = `
"import { NgModule } from '@angular/core';
import { I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { XxxModule } from \\"@spartacus/xxx\\";
import { translationChunk, translations } from \\"@spartacus/xxx/assets\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot,
    XxxModule
  ],
  providers: [provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunk,
    },
  })]
})
export class XxxFeatureModule { }
"
`;

exports[`Lib utils addLibraryFeature when the lazy loading is configured should add it in the lazy loading way 1`] = `
"import { NgModule } from '@angular/core';
import { CmsConfig, I18nConfig, provideConfig } from \\"@spartacus/core\\";
import { translationChunk, translations } from \\"@spartacus/xxx/assets\\";
import { XxxModuleRoot } from \\"@spartacus/xxx/root\\";

@NgModule({
  declarations: [],
  imports: [
    XxxModuleRoot
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      xxx: {
        module: () =>
          import('@spartacus/xxx').then((m) => m.XxxModule),
      },
    }
  }),
  provideConfig(<I18nConfig>{
    i18n: {
      resources: translations,
      chunks: translationChunk,
    },
  })
  ]
})
export class XxxFeatureModule { }
"
`;
